test/cache_test.lua

```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env tarantool
```

Comment:
I propose to test several more scenarious:

* A function in `_G` is changed without updating its name in the schema — verify that we don't cache the old callable.
* A function name in the schema is changed (does not matter, whether directly in the `_ddl_sharding_func` space or using `ddl.set_schema()`) — for the sake of completeness, we do that only for a function body for now.
* A space itself is changed (say, we add one more field into a sharding key definition) — it is why we look at the schema version.

Reply:
Done

Reply:
> A function in `_G` is changed without updating its name in the schema — verify that we don't cache the old callable.

`test_cache_func_body_changed()` does it. Okay.

> A function name in the schema is changed (does not matter, whether directly in the `_ddl_sharding_func` space or using `ddl.set_schema()`) — for the sake of completeness, we do that only for a function body for now.

`test_cache_space_updated()` does it for a stored function body. But what about updating a stored function name?

> A space itself is changed (say, we add one more field into a sharding key definition) — it is why we look at the schema version.

`test_cache_schema_changed()` looks similar on the first glance, but AFAIU it actually changes `_ddl_sharding_func` and effectively equivalent to `test_cache_space_updated()`.

I just observed that my example 'add one more field into a sharding key definition' is wrong: it is `_ddl_sharding_key` change (not a 'true' DDL) and, moreover, doesn't affect `ddl.bucket_id()` result, because it accepts a key, not a tuple.

I think more on it... and why we need to track schema version at all? The only source information for the cache is the `_ddl_sharding_func` content. We don't check even whether a space exists and maybe shouldn't (not within `ddl.bucket_id() at least).

Reply:
After voice discussion with Pavel: we should catch the situation like the following:

```lua
box.space._ddl_sharding_func:drop()
box.schema.create_space('_ddl_sharding_func', <...>)
box.space._ddl_sharding_func:create_index(<...>)
```

We can either look at the schema version (as it is implemented now) or set a trigger on `_space`.

Look at schema version: extra rebuilds at other changes, using of internal API.

Trigger on `_space`: it is not clear for me, whether we always have proper permissions to set the trigger.

So it is okay for me to keep the schema version approach. But I vote for a comment in the code that clarifies why exactly we need to look at the schema version and for rewritting the test in more clear way: so it'll be obvious that we're about the case described above.

Reply:
test added and fixed

Reply:
I'm satisfied with the test cases now.

Reply:
Thanks!